hist/hist/src/TScatter.cxx

```diff
@@ -0,0 +1,252 @@
+
+Begin_Macro(source)
+../../../tutorials/graphs/scatter.C
+End_Macro
+
+*/
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// TScatter default constructor.
+
+TScatter::TScatter(): TGraph()
+{
+   if (!CtorAllocate()) return;
+   fScale  = 5.;
+   fMargin = 0.1;
+   fSize   = nullptr;
+   fColor  = nullptr;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// TScatter normal constructor.
+///
+///  the arrays are preset to zero
+
+TScatter::TScatter(Int_t n)
+   : TGraph(n)
+{
+   if (!CtorAllocate()) return;
+   FillZero(0, fNpoints);
+   fScale  = 5.;
+   fMargin = 0.1;
+   fSize   = nullptr;
+   fColor  = nullptr;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// TScatter normal constructor.
+///
+///  if ex or ey are null, the corresponding arrays are preset to zero
+
+TScatter::TScatter(Int_t n, const Double_t *x, const Double_t *y, const Double_t *col, const Double_t *size)
+   : TGraph(n, x, y)
+{
+   if (!CtorAllocate()) return;
+
+   n = sizeof(Double_t) * fNpoints;
+   if (col) memcpy(fColor, col, n);
+   else     fColor = nullptr;
+   if (size) memcpy(fSize, size, n);
+   else      fSize = nullptr;
+   fScale  = 5.;
+   fMargin = 0.1;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// TScatter default destructor.
+
+TScatter::~TScatter()
+{
+   delete [] fColor;
+   delete [] fSize;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor allocate.
+///
+/// Note: This function should be called only from the constructor
+/// since it does not delete previously existing arrays.
+
+Bool_t TScatter::CtorAllocate()
+{
+
+   if (!fNpoints) {
+      fColor = fSize = nullptr;
+      return kFALSE;
+   } else {
+      fColor = new Double_t[fMaxSize];
+      fSize = new Double_t[fMaxSize];
+   }
+   return kTRUE;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Returns a pointer to the histogram used to draw the axis
+
+TH1F *TScatter::GetHistogram() const
+{
+//   TH1F *h;
+   if (!fHistogram) {
+      double rwxmin, rwymin, rwxmax, rwymax;
+      int npt = 100;
+      ComputeRange(rwxmin, rwymin, rwxmax, rwymax);
+      double dx = (rwxmax-rwxmin)*fMargin;
+      double dy = (rwymax-rwymin)*fMargin;
+      auto h = new TH1F(Form("%s_h",GetName()),GetTitle(),npt,rwxmin,rwxmax);
+      h->SetMinimum(rwymin-dy);
+      h->SetMaximum(rwymax+dy);
+      h->GetXaxis()->SetLimits(rwxmin-dx,rwxmax+dx);
+      h->SetBit(TH1::kNoStats);
+      h->SetDirectory(0);
+      h->Sumw2(kFALSE);
+
+      // do not add the histogram to gDirectory
+      // use local TDirectory::TContect that will set temporarly gDirectory to a nullptr and
+      // will avoid that histogram is added in the global directory
+      {
+         TDirectory::TContext ctx(nullptr);
```

Comment:
`TDirectory::TContext ctx(nullptr)` should be used with `TH1F` constructor. 
See how `TGraph::GetHistogram()` is implemented.

Reply:
Histogram constructor should be inside scope of 
```
{
   TDirectory::TContext ctx(nullptr);
   h = new TH1F(....)
}
```
Otherwise histogram will be owned by current gDirectory and can  be deleted at any time

Reply:
I changed that.

Reply:
Yes. But you do not need double braces `{ {` - the only should be ok.